### ISSUE TITLE
Minimizing over global variables instead of trainable variables

### ISSUE THREAD

**Issue by unrealwill** [openai/spinningup#5]
In https://github.com/openai/spinningup/blob/master/spinup/algos/sac/sac.py 
L185
train_pi_op = pi_optimizer.minimize(pi_loss, var_list=get_vars('main/pi'))

```
def get_vars(scope):
return [x for x in tf.global_variables() if scope in x.name]
```

you minimize over the global variables, instead of trainable variables as it should (because the network is user defined).

**Comment by machinaut** [comment]
I'm 90% sure you can include untrainable variables in the list and tensorflow does just fine.

Checking this now.

**Comment by machinaut** [comment]
@jachiam it looks like tensorflow will happily update variables with `trainable=False`.  News to me.

Filed a bug on tensorflow here https://github.com/tensorflow/tensorflow/issues/23639 but for the meantime we should probably change the scoping.

@unrealwill are you running into problems with this?  When I look at the graph all of the variables included are variables we want to train/update.

**Comment by unrealwill** [comment]
@machinaut The problem is not that we miss some variables, but that other mat be included. 
For example if I use some custom layer with non-trainable variables inside my definition of my policy, (like BatchNormalization where then the running mean and variance would also be optimized upon). 
There is currently no bug in the provided example but this is a semantic mistake which may bite the non careful users.

Also, on the same line there is a potential oher bug in get_vars('main/pi') (which is painful to me as I once got burned with a similar mistake) because imagine that you have another scope like main/piold, which you don't want to minimize on but it will get selected silently by this scope filter. Instead we should have used get_vars("main/pi/"), or even better specify scope objects. (In the code two line below the same functionality is abused with "main/q" to select both "main/q1" and "main/q2"

**Comment by unrealwill** [comment]
Other bug in the filtering
```
def get_vars(scope):
return [x for x in tf.global_variables() if scope in x.name]
```
the filtering using if scope in x.name will return the scope for other agents as well.
It is different than tf.global_variables(scope=x.name)

If you have two actor_critics one in scope agent1/main/pi and the other in scope agent2/main/pi, the code will optimize both agents at the same time.

**Comment by jachiam** [comment]
This is a valid critique. I hope you don't mind, but this one will be a bit slower to fix; my preference is to only commit changes to the implementation code if I correspondingly update the entire set of published benchmarks, and that will take a bit of time to set up and publish (and a lot more time if I wind up having to do some kind of debugging).
